Enable :expression-literals for sqlserver and oracle

appleby opened the pull request:
Closes QUE-858
Closes QUE-859

### Description

Enable the `:expression-literals` feature for sqlserver and oracle.

**Add a new `sql.qp.expression-literals/boolean->comparison` abstract driver**

This is similar in spirit to the existing `sql.qp.empty-string-is-null` driver and can be used for drivers like oracle
and sqlserver that require booleans to be converted to comparison expressions in logical operators like `:and`, `:or`, and `:not`, and in the top-level of a `:filter` clause.

In other words, oracle and sqlserver consider queries like these to be invalid

    SELECT 1 WHERE 1
    SELECT 1 WHERE 1 AND 1
    SELECT 1 WHERE 0 OR 1

But these are valid

    SELECT 1 WHERE (1 = 1)
    SELECT 1 WHERE (1 = 1) AND (1 = 1)
    SELECT 1 WHERE (0 = 1) OR  (1 = 1)

**Support selecting an empty string literal in `select-clause-deduplicate-aliases`.**

For drivers like oracle that inherit from `::sql.qp.empty-string-is-null/empty-string-is-null`, if a user adds an empty string as a custom literal expression, that literal `""` value will get compiled to `nil`, so ensure that
`select-clause-deduplicate-aliases` can handle such nil values in the `:fields` list.

Basically, we want to ensure that a query like

    {:expressions {"empty": [:value ""]},
     :fields      [[:expression "empty"]]
     :limit       1}

gets compiled to

    SELECT NULL [AS] "EMPTY" FROM ...

whereas prior to these changes, it would compile to the following

    SELECT FROM ...

### How to verify

Tests now pass for sqlserver and oracle.

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

appleby committed:
Fix literal-expressions-inside-nested-and-filtered-aggregations-test

appleby committed:
Use mt/formatted-rows in order-by-integer-literal-expression-test

appleby committed:
Support selecting an empty string literal in select-clause-deduplicate-aliases

For drivers like Oracle that inherit from ::sql.qp.empty-string-is-null/empty-string-is-null, if a user adds an empty
string as a custom literal expression, that literal "" value will get compiled to nil, so ensure that
select-clause-deduplicate-aliases can handle such nil values in the :fields list.

Basically, we want to ensure that a query like

  {:expressions {"empty": [:value ""]},
   :fields      [[:expression "empty"]]
   :limit       1}

gets compiled to

  SELECT NULL [AS] "EMPTY" ...

appleby committed:
Add sql.qp.expression-literals/boolean->comparison abstract driver

This is similar in spirit to the existing sql.qp.empty-string-is-null driver and can be used for drivers like oracle
and sqlserver that require booleans to be converted to comparison expressions in logical operators like :and, :or, and
:not, and in the top-level of a :filter clause.

appleby committed:
Enable :expression-literals for oracle

appleby committed:
Enable :expression-literals for sqlserver

metamben reviewed (approved):
LGTM, I wonder though how we can tell that this is complete. For example, do these drivers support things like `[:case true-value ...]`?

appleby committed:
Review feedback: not -> nil?

Co-authored-by: metamben <103100869+metamben@users.noreply.github.com>

appleby commented:
> LGTM, I wonder though how we can tell that this is complete. For example, do these drivers support things like `[:case true-value ...]`?

Funny you should mention it. I had the same thought about `case` this morning and so added it as well (see latest commit).

appleby committed:
Implement ->honeysql :sum-where and :distinct-where in terms of :case for :sql driver

Change the ->honeysql methods for :sum-where and :distinct-where to build an MBQL :case form and compile that, rather
than building a honeysql :case form directly. In this way, drivers that need special handling for :case will get
:sum-where and :distinct-where for free. Note that :count-where was already implemented this way. See, e.g. the
:metabase.driver.sql.query-processor.expression-literals/boolean->comparison abstract driver which converts booleans
in the case conditions to equivalent := expressions.

appleby committed:
Fix typo in test: case clause -> expression

metamben reviewed (approved):
What about `:if`?

appleby commented:
> What about `:if`?

`:if` gets desugared to `:case`.

https://github.com/metabase/metabase/blob/master/src/metabase/legacy_mbql/util.cljc#L304

appleby committed:
Rename boolean->comparison abstract driver to boolean-is-comparison

Rename

  ::sql.qp.expression-literals/boolean->comparison to
  ::sql.qp.boolean-is-comparison/boolean-is-comparison

This makes it clear that the namespace should only contain the abstract driver, and is also more consistent with the
existing driver in sql.qp.empty-string-is-null.
